tests/unit/test_charm.py

```diff
@@ -3,18 +3,55 @@
 #
 # Learn more about testing at: https://juju.is/docs/sdk/testing
 
-import unittest
-
 import ops
 import ops.testing
-from charm import OpenstackExporterOperatorCharm
+import pytest
+from charm import CLOUD_NAME, OS_CLIENT_CONFIG, SNAP_NAME, OpenstackExporterOperatorCharm
 
 
-class TestCharm(unittest.TestCase):
-    def setUp(self):
+class TestCharm:
+    def setup_method(self, _):
         self.harness = ops.testing.Harness(OpenstackExporterOperatorCharm)
-        self.addCleanup(self.harness.cleanup)
+
+    def teardown_method(self, _):
+        self.harness.cleanup()
+
+    @pytest.mark.parametrize(
+        "config",
+        [
+            ({"cache": False, "cache_ttl": "100s"}),
+            ({"cache": True, "cache_ttl": "200s"}),
+        ],
+    )
+    def test_config_changed(self, config, mocker):
+        mock_get_installed_snap_service = mocker.patch("charm.get_installed_snap_service")
+        mock_snap_service = mocker.Mock(spec_set=["configure", "restart_and_enable", "stop"])
+        mock_get_installed_snap_service.return_value = mock_snap_service
+
         self.harness.begin()
 
-    def test_null(self):
-        """Remove me."""
+        mock_get_keystone_data = mocker.Mock()
+        mock_keystone_data = mocker.Mock()
+        mock_get_keystone_data.return_value = mock_keystone_data
+        self.harness.charm._get_keystone_data = mock_get_keystone_data
```

Comment:
Checking this function in [charm.py](https://github.com/canonical/openstack-exporter-operator/blob/main/src/charm.py#L125) where it returns dictionary, so why we are using mock object here?

Reply:
I think that lines 33-36 should be changed to simply `self.harness.charm._get_keystone_data.return_value =  exp_keystone_data` and then use in line 55 `mock_write_cloud_config.assert_called_with(exp_keystone_data)`..

Reply:
Updated

Reply:
I was thinking about something like this:
```python
exp_keystone_data = {"key": "value"}
self.harness.charm._get_keystone_data.return_value =  exp_keystone_data
...
mock_write_cloud_config.assert_called_with(exp_keystone_data)
```

Reply:
Doesn't it the same? We only verify it's used as argument for `write_cloud_config`.

Reply:
I would be more understandable what we are doing, but yes you can pass eny object there even None.

Reply:
Sorry I don't think it's necessary. The Mock object has unique id, it's enough to verify what we want to test.

Reply:
I do not think this is good approach to mock thinks, but I'll not block this PR.